### DISCUSSION TITLE
Add a support for manually setting the activation of a scene as a default option

### DISCUSSION THREAD

**Discussion by IrishMerlin** [FirstGearGames/FishNet#206] [Feature Request]
I wanted to be able to load the next scene but do not activate it. Once the scene is loaded, I then fade the screen to black. Activate the scene and then un-fade the screen to normal.

I'm unable to do so with current version of FishNet SceneManager. The allowSceneActivation is automatically handled by the DefaultSceneProcessor.

It would be nice to manually set the scene activation with my script after the OnLoadEnd event is called.

**Comment by Katerlad** [comment]
To Add to this I looked into the situation IrishMerlin was in, and we looked at options in his Priority help thread. https://discord.com/channels/424284635074134018/1059879901919977523

When he mentions "Activate the Scene" he means the AsyncOperation.AllowSceneActivation bool when loading scenes. 

This suggestion is adding control of whether the scene manager should automatically activate a scene, or if there could be a load option that allows us to turn automatically activate off. 

Then have our own logic on scene load end to run, before activating it. Perfect example would be IrishMerlins example a fader that you want to fade in and out before the scene activates.

**Comment by Katerlad** [comment]
The Solution of Activating the Paused scene by the user would be the tricky part.

Almost like if a load option where to exist to turn off automatically Activating Scenes. Then when the scene is at 90% trigger an event for the scenes waiting activation. maybe something like this. 

```csharp
public void Awake()
{
SceneManager.OnAwaitingActivation += HandleAwaitingActivationScene;
SceneManager.OnLoadEnd += HandleOnLoadEnd;
}

public void HandleAwaitingActivationScene(AsyncOperation scene)
{
StartCoroutine("ActivateScene")
}

public Ienumerable ActivateScene(AsyncOperation scene)
{
// yield return StartCoroutine(Fader.ToBlack());
// unload sceneloader scene;
// scene.AllowSceneActivation = true;
}
public void HandleOnLoadEnd()
{
fader.ToClear();
}
```

**Comment by FirstGearGames** [comment]
You can do this with a custom SceneProcessor.
See DefaultSceneProcessor.

Copy, make your own, and just override the 'IsPercentComplete()' until you're happy.

**Reply by Katerlad** [reply]
Not sure if that would work, isPercentComplete() is triggered for loading and unloading for each individual scene in a queuedata.

So if you overridden this method to not return true until a fader is complete. 
1. you'd have to get reference to your fader somehow which DefaultSceneProcesser is a normal class. 
2. the fader would trigger on each unloaded scene, and on multiple loaded scenes in a single queuedata.
3. if you made an event, and triggered on event here to pass the AsyncOperation to the fader that subscribed to the event. There is no way for you to pause the IsPercentComplete to wait for an action to finish. Fishnet will auto activate the scenes once all scenes in a queuedata show IsPersentComplete() true.

I am not sure if you had something more simple in mind that I am just missing completely, but also if you are running as host, you cannot pause the servers loading and wait for a fader to complete then finish loading. Clients will be waiting for all of that to happen before they even start loading. 

I am uploading a pull request with my proposed method of implementing the feature. Its is not much code change, but does change __LoadScenes Method to Send the broadcasts to clients before activating the scenes. I have tested it out in different scenario's and it does not break anything just provides more flexablilty, however its hard to test every use case. Adding some notes into SceneManager now and will have the pull request up to be looked at here in 30 minutes. 

Let me know if you ever want to pair and screen share if you'd like to go over more extensively. I understand if its too much to take in as well, and you just dont feel comfortable adding it, but It seems like its a very good point that this functionality that is apart of unity needs to be available somehow in fishnet.

**Comment by Katerlad** [comment]
@IrishMerlin Pull Request is up with my changes if you wanted to test out my implementation, if you run into any problems. Comment on that pull request.

**Comment by FirstGearGames** [comment]
I closed the PR request, won't be taking it sorry. This can already be done with the included scene processors. I described briefly how to on your PR https://github.com/FirstGearGames/FishNet/pull/207

**Reply by Katerlad** [reply]
I probably should of replied here, I read your explanation and responded on the PR, I still don't believe it is possible to do what we are asking with the scene processor. Fishnet still controls at what point the scene activation occurs and there is no way for the user to do custom Monobehaviour logic  before scene is activated and wait for that custom monobehaviour logic to complete like a coroutine, before allowing the scenemanager to call activation. See the PR if you want to continue conversation but if not I understand haha.

**Comment by FirstGearGames** [comment]
I replied to the other one, so ignore it. Here's my response.

This is exactly what the processor does. See DefaultSceneProcessor, it's the one that FishNet defaults to but you can make your own.

https://fish-networking.gitbook.io/docs/manual/components/managers/scenemanager

**Reply by Katerlad** [reply]
Whoops scene processor is a mono-behavior how did I miss that.... I'm feeling really stupid right now.

**Comment by Katerlad** [comment]
I apologize.. im gonna retreat with my tail between my legs and not look over simple facts like that the best I can.. Wasting peeps time. but yeah That would totally work lol
